pdns-recursor-4.3.0 build broken on illumos

<!-- Hi! Thanks for filing an issue. It will be read with care by human beings. Can we ask you to please fill out this template and not simply demand new features or send in complaints? Thanks! -->
<!-- Also please search the existing issues (both open and closed) to see if your report might be duplicate -->
<!-- Please don't file an issue when you have a support question, send support questions to the mailinglist or ask them on IRC (https://www.powerdns.com/opensource.html) -->

<!-- Tell us what is issue is about -->
 - Program: Recursor
 - Issue type: Bug report

### Short description

```
portsmplexer.cc: In function 'FDMultiplexer* makePorts()':
portsmplexer.cc:43:33: error: invalid new-expression of abstract class type 'PortsFDMultiplexer'
   return new PortsFDMultiplexer();
                                 ^
portsmplexer.cc:17:7: note:   because the following virtual functions are pure within 'PortsFDMultiplexer':
 class PortsFDMultiplexer : public FDMultiplexer
       ^~~~~~~~~~~~~~~~~~
In file included from portsmplexer.cc:9:0:
mplexer.hh:84:16: note: 	virtual void FDMultiplexer::getAvailableFDs(std::vector<int>&, int)
   virtual void getAvailableFDs(std::vector<int>& fds, int timeout) = 0;
                ^~~~~~~~~~~~~~~
In file included from portsmplexer.cc:9:0:
mplexer.hh:163:23: note: 	virtual std::__cxx11::string FDMultiplexer::getName() const
   virtual std::string getName() const = 0;
                       ^~~~~~~
gmake[2]: *** [Makefile:1522: portsmplexer.o] Error 1
```

### Environment
<!-- Tell us about the environment -->
 - Operating system: SmartOS/illumos, and probably Solaris
 - Software version: 4.3.0
 - Software source: pkgsrc

### Steps to reproduce

```
cd pkgsrc/net/powerdns-recursor
bmake
```
### Expected behaviour

Build works.

### Actual behaviour

Build error log available here: https://us-east.manta.joyent.com/pkgsrc/public/reports/upstream-trunk/20200413.2254/powerdns-recursor-4.3.0/build.log

### Other information

Sorry, I don't have time to dig into this at the moment, so have just resorted to creating this bug report for now.  Hacking the configure script to disable `have_solaris` makes the build succeed, presumably using a poll backend or similar instead of event ports.

--- a/pdns/portsmplexer.cc
+++ b/pdns/portsmplexer.cc
@@ -23,11 +23,12 @@ public:
     close(d_portfd);
   }
 
-  virtual int run(struct timeval* tv, int timeout=500);
+  virtual int run(struct timeval* tv, int timeout=500) override;
+  virtual void getAvailableFDs(std::vector<int>& fds, int timeout) override;
 
-  virtual void addFD(callbackmap_t& cbmap, int fd, callbackfunc_t toDo, const boost::any& parameter, const struct timeval* ttd=nullptr);
-  virtual void removeFD(callbackmap_t& cbmap, int fd);
-  string getName()
+  virtual void addFD(callbackmap_t& cbmap, int fd, callbackfunc_t toDo, const boost::any& parameter, const struct timeval* ttd=nullptr) override;
+  virtual void removeFD(callbackmap_t& cbmap, int fd) override;
+  string getName() const override
   {
     return "solaris completion ports";
   }
@@ -78,6 +79,58 @@ void PortsFDMultiplexer::removeFD(callbackmap_t& cbmap, int fd)
     throw FDMultiplexerException("Removing fd from port set: "+stringerror());
 }
 
+void PortsFDMultiplexer::getAvailableFDs(std::vector<int>& fds, int timeout)
+{
+  struct timespec timeoutspec;
+  timeoutspec.tv_sec = timeout / 1000;
+  timeoutspec.tv_nsec = (timeout % 1000) * 1000000;
+  unsigned int numevents = 1;
+  int ret = port_getn(d_portfd, d_pevents.get(), min(PORT_MAX_LIST, s_maxevents), &numevents, &timeoutspec);
+
+  /* port_getn has an unusual API - (ret == -1, errno == ETIME) can
+     mean partial success; you must check (*numevents) in this case
+     and process anything in there, otherwise you'll never see any
+     events from that object again. We don't care about pure timeouts
+     (ret == -1, errno == ETIME, *numevents == 0) so we don't bother
+     with that case. */
+  if (ret == -1 && errno != ETIME) {
+    if (errno != EINTR) {
+      throw FDMultiplexerException("completion port_getn returned error: " + stringerror());
+    }
+
+    // EINTR is not really an error
+    return;
+  }
+
+  if (numevents == 0) {
+    // nothing
+    return;
+  }
+
+  fds.reserve(numevents);
+
+  for (unsigned int n = 0; n < numevents; ++n) {
+    const auto fd = d_pevents[n].portev_object;
+
+    /* we need to re-associate the FD */
+    if (d_readCallbacks.count(fd)) {
+      if (port_associate(d_portfd, PORT_SOURCE_FD, fd, POLLIN, 0) < 0) {
+        throw FDMultiplexerException("Unable to add fd back to ports (read): " + stringerror());
+      }
+    }
+    else if (d_writeCallbacks.count(fd)) {
+      if (port_associate(d_portfd, PORT_SOURCE_FD, fd, POLLOUT, 0) < 0) {
+        throw FDMultiplexerException("Unable to add fd back to ports (write): " + stringerror());
+      }
+    } else {
+      /* not registered, this is unexpected */
+      continue;
+    }
+
+    fds.push_back(fd);
+  }
+}
+
 int PortsFDMultiplexer::run(struct timeval* now, int timeout)
 {
   if(d_inrun) {
@@ -85,8 +138,8 @@ int PortsFDMultiplexer::run(struct timeval* now, int timeout)
   }
   
   struct timespec timeoutspec;
-  timeoutspec.tv_sec = time / 1000;
-  timeoutspec.tv_nsec = (time % 1000) * 1000000;
+  timeoutspec.tv_sec = timeout / 1000;
+  timeoutspec.tv_nsec = (timeout % 1000) * 1000000;
   unsigned int numevents=1;
   int ret= port_getn(d_portfd, d_pevents.get(), min(PORT_MAX_LIST, s_maxevents), &numevents, &timeoutspec);
   

--- a/pdns/test-dns_random_hh.cc
+++ b/pdns/test-dns_random_hh.cc
@@ -123,7 +123,7 @@ BOOST_AUTO_TEST_CASE(test_dns_random_getrandom_average) {
 #endif
 
 #if defined(HAVE_ARC4RANDOM)
-BOOST_AUTO_TEST_CASE(test_dns_random_getrandom_average) {
+BOOST_AUTO_TEST_CASE(test_dns_random_arc4random_average) {
 
   ::arg().set("rng")="arc4random";
   ::arg().set("entropy-source")="/dev/urandom";

